src/RecurringPurchase.sol

```diff
@@ -0,0 +1,270 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.23;
+
+import {IERC20} from "openzeppelin/token/ERC20/IERC20.sol";
+import {SafeERC20} from "openzeppelin/token/ERC20/utils/SafeERC20.sol";
+import {IERC20Metadata} from "openzeppelin/token/ERC20/extensions/IERC20Metadata.sol";
+
+import {AggregatorV3Interface} from "src/vendor/chainlink/AggregatorV3Interface.sol";
+import {ISwapRouter} from "v3-periphery/interfaces/ISwapRouter.sol";
+
+import {QuarkWallet} from "quark-core/src/QuarkWallet.sol";
+import {QuarkScript} from "quark-core/src/QuarkScript.sol";
+
+/**
+ * @title Recurring Purchase Script
+ * @notice Quark script that performs a swap on a regular interval.
+ * @author Legend Labs, Inc.
+ */
+contract RecurringPurchase is QuarkScript {
+    using SafeERC20 for IERC20;
+
+    error BadPrice();
+    error InvalidInput();
+    error PurchaseConditionNotMet();
+
+    /// @notice Emitted when a swap is executed
+    event SwapExecuted(
+        address indexed sender,
+        address indexed recipient,
+        address tokenIn,
+        address tokenOut,
+        uint256 amountIn,
+        uint256 amountOut,
+        bytes path
+    );
+
+    /// @notice The base slippage factor where `1e18` represents 100% slippage tolerance
+    uint256 public constant BASE_SLIPPAGE_FACTOR = 1e18;
+
+    /**
+     * @dev Note: This script uses the following storage layout in the QuarkStateManager:
+     *         mapping(bytes32 hashedPurchaseConfig => uint256 nextPurchaseTime)
+     *             where hashedPurchaseConfig = keccak256(PurchaseConfig)
+     */
+
+    /// @notice Parameters for a recurring purchase order
+    struct PurchaseConfig {
+        uint256 interval;
+        SwapParams swapParams;
+        SlippageParams slippageParams;
+    }
+
+    /// @notice Parameters for performing a swap
+    struct SwapParams {
+        address uniswapRouter;
+        address recipient;
+        address tokenIn;
+        address tokenOut;
+        /// @dev The amount for tokenIn if exact in; the amount for tokenOut if exact out
+        uint256 amount;
+        /// @dev False for exact in; true for exact out
+        bool isExactOut;
+        uint256 deadline;
+        bytes path;
+    }
+
+    /// @notice Parameters for controlling slippage in a swap operation
+    struct SlippageParams {
+        /// @dev Maximum acceptable slippage, expressed as a percentage where 100% = 1e18
+        uint256 maxSlippage;
+        /// @dev Price feed addresses for determining market exchange rates between token pairs
+        /// Example: For SUSHI -> SNX swap, use [SUSHI/ETH feed, SNX/ETH feed]
+        address[] priceFeeds;
+        /// @dev Flags indicating whether each corresponding price feed should be inverted
+        /// Example: For USDC -> ETH swap, use [true] with [ETH/USD feed] to get ETH per USDC
+        bool[] shouldReverses;
```

Comment:
nit: should we do a struct of array instead?

Reply:
Yeah, an array of structs could work as well. It's kind of a trade-off because we'll introduce another layer of nested structs, which is annoying to deal with in solidity. So I'm not completely sold on either approach, but can be swayed if someone has a stronger opinion on it.

Reply:
Prior experience lands me on the side of "struct-of-arrays". So I would leave this as-is. :+1: